sale_substate/models/sale_order.py

```diff
@@ -0,0 +1,51 @@
+# Copyright 2019 Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class BaseSubstateType(models.Model):
+    _inherit = "base.substate.type"
+
+    model = fields.Selection(selection_add=[("sale.order", "Sale order")])
+
+
+class SaleOrder(models.Model):
+    _inherit = ["sale.order", "base.substate.mixin"]
+    _name = "sale.order"
+
+    @api.constrains("substate_id", "state")
+    def check_substate_id_value(self):
```

Comment:
Can this function be included in `base.substate.mixin` using technique like this,
https://github.com/OCA/account-analytic/blob/12.0/analytic_tag_dimension_enhanced/models/analytic.py#L77
If this method and `_track_template()` can be in base module, this would be awesome. Not sure you there are technical reason not to do so.

Note: I am not so sure, it is applicable for api.constrains.